### PR TITLE
[improvements] Return error to application and support SLO request validation

### PR DESCRIPTION
**Pull request by Zogoo**
This PR contains the following improvements. 
The reason for combining these changes into a single pull request (PR) is that they do not directly affect the gem's core features. Splitting my code into smaller PRs would require manual tasks, which can introduce bugs. Additionally, all these improvements have already been tested in a real environment.

1. Currently, the gem is failing silently and only provides logs which makes the application cannot take any actions based on failure. For example, it should display errors related to the customer's configurations. To address this, I added an error collection to the SAML request. After validation, these errors will be stored in an array for further processing.
2. Allow the application to retrieve metadata for usage that is not defined as `unspecified_certificate`.
3. Follow the metadata instructions on how to validate the Authn Request. If the metadata specifies that it should be signed, the IdP should accept a signed Authn Request.

### CODE REVIEW

File: lib/saml_idp/controller.rb [hunk 1]
```diff
@@ -34,10 +34,7 @@ def acs_url
 
     def validate_saml_request(raw_saml_request = params[:SAMLRequest])
       decode_request(raw_saml_request, params[:Signature], params[:SigAlg], params[:RelayState])
-      return true if valid_saml_request?
-
-      head :forbidden if defined?(::Rails)
```
Comment (Zogoo) [inline lib/saml_idp/controller.rb:39]: This was not UX friendly option that rendered a black screen from the gem. We should leave this decision to the application side and only provide helper methods.

**Comment by Zogoo** [comment]
@jphenow can you look at this one?

File: lib/saml_idp/incoming_metadata.rb [hunk 2]
```diff
@@ -63,6 +63,15 @@ def contact_person
     end
     hashable :contact_person
 
+    def unspecified_certificate
```
Comment (Zogoo) [inline lib/saml_idp/incoming_metadata.rb:66]: One example of SAML 2.0 spec it has too many optional fields for a too generic solution.
Gem missing if there certificate without specifying usage (`@use='signing'`) of the certificate. For this reason, we need to at least make it retrievable for application.

File: lib/saml_idp/request.rb [hunk 3]
```diff
@@ -3,6 +3,8 @@
 require 'logger'
 module SamlIdp
   class Request
+    attr_accessor :errors
```
Comment (Zogoo) [inline lib/saml_idp/request.rb:6]: Collecting errors for application then application can decide action based on error.

File: lib/saml_idp/request.rb [hunk 4]
```diff
@@ -151,7 +170,14 @@ def valid_external_signature?
         OpenSSL::Digest::SHA1
       end
 
-      cert.public_key.verify(signature_algorithm.new, raw_signature, query_request_string)
+      result = cert.public_key.verify(signature_algorithm.new, raw_signature, query_request_string)
+      # Match all percent-encoded sequences (e.g., %20, %2B) and convert them to lowercase
+      # Upper case is recommended for consistency but some services such as MS Entra Id not follows it
+      # https://datatracker.ietf.org/doc/html/rfc3986#section-2.1
+      result || cert.public_key.verify(signature_algorithm.new, raw_signature, query_request_string.gsub(/%[A-F0-9]{2}/) { |match| match.downcase })
```
Comment (Zogoo) [inline lib/saml_idp/request.rb:177]: Lowercase URL encoding is also possible, but the signature is case-sensitive. 
For this case, we need to try both cases for signature verification.

File: spec/lib/saml_idp/request_spec.rb
Comment (Zogoo) [inline spec/lib/saml_idp/request_spec.rb:1]: Proper SLO test with ruby-saml gem.

File: spec/support/security_helpers.rb [hunk 5]
```diff
@@ -51,7 +51,7 @@ def signature_fingerprint_1
     @signature_fingerprint1 ||= "C5:19:85:D9:47:F1:BE:57:08:20:25:05:08:46:EB:27:F6:CA:B7:83"
   end
 
-  def signature_1
+  def certificate_1
```
Comment (Zogoo) [inline spec/support/security_helpers.rb:54]: It is certificate not a signature.

**Review by jphenow — APPROVED** [review]

**Pull request merged**